### PR TITLE
logging: Refactor log_output module

### PR DESCRIPTION
**Pull request by nordic-krch**
Module refactored: splitted data into read-only part and control block,
adding macro for creating log_output instance

There are no functional changes in the module rather cleaning up initial implementation.
Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

### CODE REVIEW

File: subsys/logging/log_backend_uart.c [hunk 1]
```diff
@@ -45,15 +40,18 @@ static void put(const struct log_backend *const backend,
...
 
 }
 
 void log_backend_uart_init(void)
 {
-	ctx.ctx = device_get_binding(CONFIG_UART_CONSOLE_ON_DEV_NAME);
+	struct device *dev;
+
+	dev = device_get_binding(CONFIG_UART_CONSOLE_ON_DEV_NAME);
+	log_output_ctx_set(&log_output, dev);
```
Comment (jarz-nordic) [inline subsys/logging/log_backend_uart.c]: maybe assert if binding not success ?

File: subsys/logging/log_output.c [hunk 2]
```diff
@@ -43,53 +43,55 @@ static u32_t timestamp_div;
...
 	va_end(args);
 
 	return length;
 }
 
-static void flush(struct log_output_ctx *ctx)
+void log_output_flush(const struct log_output *log_output)
 {
-	ctx->func(ctx->data, ctx->offset, ctx->ctx);
+	log_output->func(log_output->buf, log_output->control_block->offset,
```
Comment (pizi-nordic) [inline subsys/logging/log_output.c]: Could flush function fail of do only partial flush?

File: subsys/logging/log_output.c [hunk 3]
```diff
@@ -250,7 +252,7 @@ static void std_print(struct log_msg *msg,
 }
```
Comment (pizi-nordic) [inline subsys/logging/log_output.c]: Please add default: and __ASSERT() catching unsupported number of arguments.

File: subsys/logging/log_output.c [hunk 4]
```diff
@@ -310,25 +313,26 @@ static void hexdump_print(struct log_msg *msg,
...
+	assert(log_output->size);
 
 	size_t offset = 0;
 	size_t length;
 
 	do {
-		length = ctx->length;
-		log_msg_hexdump_data_get(msg, ctx->data, &length, offset);
+		length = log_output->size;
+		log_msg_hexdump_data_get(msg, log_output->buf, &length, offset);
```
Comment (pizi-nordic) [inline subsys/logging/log_output.c]: s/log_msg_hexdump_data_get/log_msg_raw_data_get/ ?
It is a bit confusing using hexdump to access string.

File: subsys/logging/log_output.c [hunk 5]
```diff
@@ -292,14 +295,14 @@ static u32_t hexdump_line_print(struct log_msg *msg,
 }
 
 static void hexdump_print(struct log_msg *msg,
-			  struct log_output_ctx *ctx,
+			  const struct log_output *log_output,
 			  int prefix_offset)
```
Comment (pizi-nordic) [inline subsys/logging/log_output.c]: Where are you handling description string embedded in hexdump message?

File: include/logging/log_output.h [hunk 6]
```diff
@@ -36,29 +36,69 @@ extern "C" {
...
- * @param ctx Context.
  * @param flags Optional flags.
  */
-void log_output_msg_process(struct log_msg *msg,
-			    struct log_output_ctx *ctx,
+void log_output_msg_process(const struct log_output *log_output,
+			    struct log_msg *msg,
 			    u32_t flags);
 
+/** @brief Function for flushing output buffer.
```
Comment (pfalcon) [inline include/logging/log_output.h]: Nitpick, but these "Function for" intros look superfluous. I guess, they won't look pretty in the generated docs. I'd suggest to remove them. "Flush output buffer." is good enough @brief description. @dbkinder: Any suggestions here?

**Comment by pfalcon** [comment]
There's a merge conflict here now and apparently unaddressed comments. Is this targeted for 1.13? I guess it makes sense to, so 1.13 had new logging in the shape its authors want it too. This patch is also included in https://github.com/zephyrproject-rtos/zephyr/pull/9362 and complicates review there.

File: subsys/logging/log_backend_uart.c (hunk 1 above)
Comment (nordic-krch) [reply subsys/logging/log_backend_uart.c]: added

File: subsys/logging/log_output.c [hunk 7]
```diff
@@ -43,53 +43,55 @@ static u32_t timestamp_div;
...
 			   bool format)
 {
 	int length;
 	u32_t timestamp = log_msg_timestamp_get(msg);
 
 	if (!format) {
-		length = print(ctx, "[%08lu] ", timestamp);
+		length = print(log_output, "[%08lu] ", timestamp);
 	} else if (freq) {
 		u32_t reminder;
```
[the remark this answers is not in the document]
Comment (nordic-krch) [reply subsys/logging/log_output.c]: fixed

File: subsys/logging/log_output.c [hunk 8]
```diff
@@ -43,53 +43,55 @@ static u32_t timestamp_div;
...
 
-	out_ctx->data[out_ctx->offset] = (u8_t)c;
-	out_ctx->offset++;
+	out_ctx->buf[out_ctx->control_block->offset] = (u8_t)c;
+	out_ctx->control_block->offset++;
 
-	if (out_ctx->offset == out_ctx->length) {
-		out_ctx->func(out_ctx->data, out_ctx->length, out_ctx->ctx);
-		out_ctx->offset = 0;
+	if (out_ctx->control_block->offset == out_ctx->size) {
```
[the remark this answers is not in the document]
Comment (nordic-krch) [reply subsys/logging/log_output.c]: added

File: subsys/logging/log_output.c [hunk 9]
```diff
@@ -43,53 +43,55 @@ static u32_t timestamp_div;
...
+		log_output_flush(out_ctx);
 	}
 
 	return 0;
 }
 
-static int print(struct log_output_ctx *ctx, const char *fmt, ...)
+static int print(const struct log_output *log_output, const char *fmt, ...)
```
[the remark this answers is not in the document]
Comment (nordic-krch) [reply subsys/logging/log_output.c]: renamed

File: subsys/logging/log_output.c (hunk 2 above)
Comment (nordic-krch) [reply subsys/logging/log_output.c]: i've rewritten this function to take into account value returned by log_output->func()

File: subsys/logging/log_output.c (hunk 3 above)
Comment (nordic-krch) [reply subsys/logging/log_output.c]: added

File: subsys/logging/log_output.c (hunk 4 above)
Comment (nordic-krch) [reply subsys/logging/log_output.c]: added comment.

File: subsys/logging/log_output.c (hunk 5 above)
Comment (nordic-krch) [reply subsys/logging/log_output.c]: rebased to master and now it is there.

**Comment by nordic-krch** [comment]
@pfalcon i was on 3 weeks vacation but now i'm back! Rebased and applied review comments.

**Review by pizi-nordic — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]
LGTM

**Comment by carlescufi** [comment]
@pfalcon @jarz-nordic are you happy with this?

**Review by nashif — APPROVED** [review]

**Comment by pfalcon** [comment]
There's no -1 from me, so apparently I am ;-). (I don't have time for detailed review now, trust other people's opinion). Thanks for ping.

**Pull request merged**